### PR TITLE
stat: preserve non-UTF-8 filenames in error messages

### LINKED ISSUE #14685: bug(stat): invalid unicode file not escaped, they replaced with U+FFFD
Hi, uutils mainteners

when trying

```rust
relunsec@relunsec:~/software/coreutils/src/uu/touch/src$ ../../../../target/debug/stat $'\xC3'
stat: cannot statx '�': Invalid or incomplete multibyte or wide character
relunsec@relunsec:~/software/coreutils/src/uu/touch/src$ gnustat $'\xC3'
gnustat: cannot statx ''$'\303': Invalid or incomplete multibyte or wide character
relunsec@relunsec:~/software/coreutils/src/uu/touch/src$ 
```

### PR DESCRIPTION
**Pull request by ddy314**
Preserve filenames in stat errors without lossy UTF-8 conversion, using the existing locale-aware shell escaping. The diagnostics now use GNU-compatible octal byte escapes. Fixes #14685.

Also fix the shared shell quoter when a filename contains both a single quote and invalid bytes: dollar-quoted escapes require single-quote segments. Add regression coverage and explicitly gate the byte-based stat test to Unix.

Validation: 42 stat tests, 14 quoting tests, and 350 related utility integration tests passed (2 ignored); clippy and rustfmt passed. Thirty stderr/exit-code comparisons against GNU stat 9.11 matched across C/C.UTF-8, default/-L/-f and five filename cases. The full selected-feature test command also passes with the host NO_COLOR setting removed for the color tests.

AI assistance: Codex helped implement, test, and respond to review under the account owner's direction.

### CODE REVIEW

File: tests/by-util/test_stat.rs [hunk 1]
```diff
@@ -881,3 +881,18 @@ stat: '%.3': invalid directive
...
+    use std::ffi::OsStr;
+    use std::os::unix::ffi::OsStrExt;
+
+    let name = OsStr::from_bytes(b"missing-\xff");
+    for args in [vec![], vec!["-L"], vec!["-f"]] {
+        new_ucmd!()
+            .args(&args)
+            .arg(name)
+            .fails_with_code(1)
+            .stderr_contains("$'missing-\\xFF'");
```
Comment (cakebaker) [inline tests/by-util/test_stat.rs]: I think the expected output is not correct. With GNU `stat` I get the following output:
```
$ stat $'missing-\xFF'
stat: cannot statx 'missing-'$'\377': No such file or directory
```

Comment (ddy314) [reply tests/by-util/test_stat.rs]: You are right: preserving the byte with Quotable still gave a different format. b36f315f6 switches these errors to the existing locale-aware shell quoter and expects octal escapes. I also fixed its single-quote/invalid-byte edge case. Thirty full stderr and exit-code comparisons against GNU stat 9.11 match across C/C.UTF-8, default/-L/-f and five filename cases. The 42 stat tests, 14 quoting tests and 350 related utility tests pass (2 ignored).

File: tests/by-util/test_stat.rs [hunk 2]
```diff
@@ -881,3 +881,18 @@ stat: '%.3': invalid directive
...
+    }
+}
```
[the remark this answers is not in the document]
Comment (ddy314) [reply tests/by-util/test_stat.rs:905]: Added an explicit #[cfg(unix)] guard to the byte-based regression in b36f315f6.

File: tests/by-util/test_stat.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ddy314) [reply tests/by-util/test_stat.rs]: Kept an exact assertion because GNU diagnostic compatibility is part of the expected behavior, as the maintainer noted below. The expectation now uses GNU-compatible octal escaping, verified against the GNU binary in C and C.UTF-8 locales.

File: src/uucore/src/lib/features/quoting_style/shell_quoter.rs [hunk 3]
```diff
@@ -107,8 +107,22 @@ pub(super) struct EscapedShellQuoter<'a> {
...
 
 impl<'a> EscapedShellQuoter<'a> {
-    pub fn new(reference: &'a [u8], always_quote: bool, dirname: bool, size_hint: usize) -> Self {
-        let (quotes, must_quote) = initial_quoting(reference, dirname, always_quote, true);
+    pub fn new(
+        reference: &'a [u8],
+        always_quote: bool,
+        dirname: bool,
+        size_hint: usize,
+        encoding: crate::i18n::UEncoding,
```
Comment (sylvestre) [inline src/uucore/src/lib/features/quoting_style/shell_quoter.rs]: please import UEncoding instead of the full path 3 times

File: src/uucore/src/lib/features/quoting_style/shell_quoter.rs [hunk 4]
```diff
@@ -107,8 +107,22 @@ pub(super) struct EscapedShellQuoter<'a> {
...
+    ) -> Self {
+        let (mut quotes, must_quote) = initial_quoting(reference, dirname, always_quote, true);
+        let invalid = match encoding {
```
Comment (sylvestre) [inline src/uucore/src/lib/features/quoting_style/shell_quoter.rs]: this rescans every name (ls calls this per entry). could you only do it when initial_quoting picked Quotes::Double?

File: src/uucore/src/lib/features/quoting_style/mod.rs [hunk 5]
```diff
@@ -338,6 +339,21 @@ mod tests {
         check_names_encoding_raw(UEncoding::Ascii, name, map);
     }
 
+    #[test]
+    fn test_invalid_bytes_with_single_quote() {
```
Comment (sylvestre) [inline src/uucore/src/lib/features/quoting_style/mod.rs:343]: could use check_names_raw_both like the other tests here, no?

File: src/uucore/src/lib/features/quoting_style/shell_quoter.rs (hunk 3 above)
Comment (ddy314) [reply src/uucore/src/lib/features/quoting_style/shell_quoter.rs]: Imported UEncoding and use the short name throughout in d6690b89b.

File: src/uucore/src/lib/features/quoting_style/shell_quoter.rs (hunk 4 above)
Comment (ddy314) [reply src/uucore/src/lib/features/quoting_style/shell_quoter.rs]: The encoding validity check now only runs when initial_quoting selected double quotes in d6690b89b.

File: src/uucore/src/lib/features/quoting_style/mod.rs (hunk 5 above)
Comment (ddy314) [reply src/uucore/src/lib/features/quoting_style/mod.rs:343]: Converted this regression to check_names_raw_both in d6690b89b.
